File: python/evm_state/source.py
"""Bind checkpoint inputs to a prepared, locally owned native ingestion run."""
from contextlib import contextmanager
import hashlib
import json
from pathlib import Path
import re
import socket

from .files import file_lock
from .proof import VerificationError, address


def _selected(values):
    if not isinstance(values, list) or not values:
        raise VerificationError("source accounts must be a nonempty list")
    return sorted({address(value) for value in values})


def _validate(client, source, identity, run_id, directory):
    if (identity.get("format_version") != 2 or identity.get("host") != socket.gethostname()
            or identity.get("state_directory") != str(directory)):
        raise VerificationError("source run has no matching host/directory binding; use a fresh guarded continuation")
    if (identity.get("database") != client.database or identity.get("http_url") != client.url
            or identity.get("module") != "map_block_state" or identity.get("network") != "bsc"
            or identity.get("schema_version") != 1 or identity.get("final_blocks_only") is not True):
        raise VerificationError("source database identity is not the qualified finalized BSC module")
    if (source.get("final_blocks_only") is not True or source.get("module_hash") != identity["module_hash"]
            or not re.fullmatch(r"[0-9a-f]{40}", identity["module_hash"])
            or _selected(source["accounts"]) != identity["accounts"]):
        raise VerificationError("source filter, module hash or finality differs from its native run")
    start = source["start_block"]
    if isinstance(start, bool) or not isinstance(start, int) or start < identity["start_block"]:
        raise VerificationError("source range starts before its native run")
    record = json.loads((directory / "run.json").read_text())
    db_uuid = client.one("SELECT toString(uuid) AS id FROM system.databases WHERE name={db:String}",
                         {"db": client.database})["id"]
    if (record.get("phase") != "prepared" or record.get("run_id") != run_id
            or record.get("identity") != identity or record.get("database_uuid") != db_uuid):
        raise VerificationError("source database and local run metadata differ")
    if hashlib.sha256((directory / "package.spkg").read_bytes()).hexdigest() != identity["package_sha256"]:
        raise VerificationError("source frozen package is missing or changed")
    metadata = directory / "meta" / f"{client.database}_schema_hash.txt"
    if metadata.read_text().strip() != record["schema_hash"]:
        raise VerificationError("source schema metadata is missing or changed")
    verified = {"run_id": run_id, "database_uuid": db_uuid, "package_sha256": identity["package_sha256"],
                "schema_hash": record["schema_hash"], "state_directory": str(directory)}
    if any(key in source and source[key] != value for key, value in verified.items()):
        raise VerificationError("source provenance differs from its recorded native run")
    return {**source, **verified}


@contextmanager
def verified_source(client, source):
    exists = int(client.one("SELECT count() AS n FROM system.tables WHERE database={db:String} "
        "AND name='_evm_state_run'", {"db": client.database})["n"])
    if not exists:
        raise VerificationError("source has no guarded native run ownership record")
    owner = client.one("SELECT run_id,identity FROM _evm_state_run")
    try:
        identity = json.loads(owner["identity"])
        # Old prototype sources can be exported through existing ready checkpoints;
        # do not silently adopt their mutable history into newly published state.
        if not isinstance(identity, dict) or identity.get("format_version") != 2:
            raise VerificationError("source uses an old native run identity; use a fresh guarded continuation")
        directory = Path(identity["state_directory"]).resolve()
        if (identity["state_directory"] != str(directory) or not directory.is_dir()
                or identity.get("host") != socket.gethostname()):
            raise VerificationError("source run has no matching host/directory binding")
    except (KeyError, TypeError, OSError, AttributeError) as error:
        raise VerificationError("source run metadata is missing or invalid") from error
    # Source-history cleanup must take this lock exclusively, in addition to
    # excluding its native writer. Appends can continue beyond our fixed end.
    with file_lock(directory / "source_readers.lock"):
        try:
            checked = _validate(client, source, identity, owner["run_id"], directory)
        except (KeyError, TypeError, OSError, AttributeError) as error:
            raise VerificationError("source run metadata is missing or invalid") from error
        yield checked

File: crates/evm-state/src/source.rs
//! Bind source inputs to a prepared locally owned native ingestion run.
use crate::{
    ch::{params, uint, ClickHouse},
    checkpoint::canonical_accounts,
    cursor,
    files::{file_lock, resolve, Lock},
    host,
    proof::string,
};
use anyhow::{ensure, Context, Result};
use serde_json::{json, Value};
use sha2::{Digest, Sha256};
use std::{fs, path::Path};

pub struct VerifiedSource {
    pub source: Value,
    _reader: Lock,
}

fn selected(value: &Value) -> Result<Vec<String>> {
    ensure!(value.is_array(), "source accounts must be a nonempty list");
    canonical_accounts(value)
}

fn validate(
    client: &ClickHouse,
    source: &Value,
    identity: &Value,
    run_id: &str,
    directory: &Path,
) -> Result<Value> {
    let path = directory.to_str().context("invalid source directory")?;
    ensure!(
        identity["format_version"] == 3
            && host::matches(identity, directory)?
            && identity["state_directory"] == path,
        "source run has no matching host/directory binding; use a fresh guarded continuation"
    );
    ensure!(
        identity["database"] == client.database
            && identity["http_url"] == client.url
            && identity["module"] == "map_block_state"
            && identity["network"] == "bsc"
            && identity["schema_version"] == 1
            && identity["final_blocks_only"] == true,
        "source database identity is not the qualified finalized BSC module"
    );
    let module = string(identity, "module_hash")?;
    ensure!(
        source["final_blocks_only"] == true
            && source["module_hash"] == module
            && module.len() == 40
            && module
                .bytes()
                .all(|v| v.is_ascii_digit() || (b'a'..=b'f').contains(&v))
            && serde_json::to_value(selected(&source["accounts"])?)? == identity["accounts"],
        "source filter, module hash or finality differs from its native run"
    );
    ensure!(
        source["start_block"]
            .as_u64()
            .context("invalid source start block")?
            >= uint(&identity["start_block"])?,
        "source range starts before its native run"
    );
    let record: Value = serde_json::from_slice(&fs::read(directory.join("run.json"))?)?;
    let db = client.one(
        "SELECT toString(uuid) AS id FROM system.databases WHERE name={db:String}",
        &params(json!({"db":client.database}))?,
    )?;
    ensure!(
        record["phase"] == "prepared"
            && record["run_id"] == run_id
            && &record["identity"] == identity
            && record["database_uuid"] == db["id"],
        "source database and local run metadata differ"
    );
    ensure!(
        hex::encode(Sha256::digest(fs::read(directory.join("package.spkg"))?))
            == string(identity, "package_sha256")?,
        "source frozen package is missing or changed"
    );
    ensure!(
        fs::read_to_string(
            directory
                .join("meta")
                .join(format!("{}_schema_hash.txt", client.database))
        )?
        .trim()
            == string(&record, "schema_hash")?,
        "source schema metadata is missing or changed"
    );
    let verified = json!({"run_id":run_id,"database_uuid":db["id"],"package_sha256":identity["package_sha256"],"schema_hash":record["schema_hash"],"state_directory":path});
    let mut checked = source.as_object().cloned().context("invalid source")?;
    for (key, value) in verified.as_object().unwrap() {
        ensure!(
            source.get(key).is_none_or(|actual| actual == value),
            "source provenance differs from its recorded native run"
        );
        checked.insert(key.clone(), value.clone());
    }
    Ok(Value::Object(checked))
}

pub fn verified_source(
    client: &ClickHouse,
    source: &Value,
    checkpoints: Option<&ClickHouse>,
    end_block: Option<u64>,
) -> Result<VerifiedSource> {
    ensure!(uint(&client.one("SELECT count() AS n FROM system.tables WHERE database={db:String} AND name='_evm_state_run'", &params(json!({"db":client.database}))?)?["n"])? != 0, "source has no guarded native run ownership record");
    let owner = client.one(
        "SELECT run_id,identity FROM _evm_state_run",
        &Default::default(),
    )?;
    let identity: Value = serde_json::from_str(string(&owner, "identity")?)?;
    ensure!(
        identity.is_object() && identity["format_version"] == 3,
        "source uses an old native run identity; use a fresh guarded continuation"
    );
    if let Some(checkpoints) = checkpoints {
        ensure!(
            identity["checkpoint_database"] == checkpoints.database,
            "source is bound to a different checkpoint database"
        );
    }
    let directory = resolve(Path::new(string(&identity, "state_directory")?))?;
    ensure!(
        identity["state_directory"] == directory.to_str().context("invalid source directory")?
            && directory.is_dir()
            && host::matches(&identity, &directory)?,
        "source run has no matching host/directory binding"
    );
    let reader = file_lock(&directory.join("source_readers.lock"), false, true)?;
    let mut checked = validate(
        client,
        source,
        &identity,
        string(&owner, "run_id")?,
        &directory,
    )?;
    if let Some(end) = end_block {
        let run = serde_json::from_slice(&fs::read(directory.join("run.json"))?)?;
        let progress = cursor::load_progress(client, &run, &directory)?;
        ensure!(
            uint(&progress["position"]["block"]["number"])? >= end,
            "durable native progress has not reached the checkpoint target"
        );
        checked["durable_position"] = progress["position"]["block"].clone();
    }
    Ok(VerifiedSource {
        source: checked,
        _reader: reader,
    })
}
